[menu] Add optional allocator for ui menu classes

MatthewMArnold opened the pull request:
This allows us to in-place allocate menus (whereas previously we were forced to use `new`). The main change I had to make was add an `IAbstractView` that isn't templated that is a parent of the `AbstractView` class to avoid circular template parameter dependencies. I'm not sure if there is a better way to do this.

MatthewMArnold committed:
[menu] Add optional allocator for ui menu classes, used for destroying AbstractViews

annawng commented on src/modm/ui/menu/abstract_menu.hpp:
```diff
@@ -30,12 +30,17 @@ namespace modm{
 	 * \ingroup modm_ui_menu
 	 */
-	class AbstractMenu: public AbstractView
+	template<typename Allocator = allocator::Dynamic<IAbstractView> >
```
I might be misunderstanding this but why do you need a template here? Is this sort of like using `Allocator` as an alias for `allocator::Dynamic<IAbstractView>`?

MatthewMArnold replied:
This is a templated class. So it allows us to specify the type of `Allocator` we want the class to use at compile time. The part `Allocator = allocator::Dynamic<IAbstractView>` is a default template parameter. If no template parameter is supplied the object will use `allocator::Dynamic<IAbstractView>` for the allocator.

annawng commented on src/modm/ui/menu/choice_menu.hpp:
```diff
@@ -37,13 +38,14 @@ namespace modm{
 	 */
-	class  ChoiceMenu: public AbstractMenu
+	template<typename Allocator = allocator::Dynamic<IAbstractView> >
+	class  ChoiceMenu: public AbstractMenu<Allocator>
```
```suggestion
	class ChoiceMenu : public AbstractMenu<Allocator>
```

annawng commented on src/modm/ui/menu/choice_menu.hpp:
```diff
@@ -112,4 +114,6 @@ namespace modm{
 	};
 }
 
+#include "choice_menu_impl.hpp"
```
Why is this being included at the end of the file?

MatthewMArnold replied:
Doing this is a way of creating a template class with separate files for declaration/definition. Since you can't put template stuff in a `.cpp` file you put it in a `_impl.hpp` file and then include it at the bottom of the normal `.hpp` file. Then you can add a preprocessor statement at the top of the `_impl.hpp` file that makes it so your code won't compile if you try and include the `_impl.hpp` file in your own code (since you should be including the `.hpp` file instead).

annawng commented on src/modm/ui/menu/choice_menu_impl.hpp:
```diff
@@ -11,43 +12,48 @@
 	position(0)
 {
-	this->maximalDrawnEntrys = (getViewStack()->getDisplay().getHeight()- 16) / 8 ;
+	this->maximalDrawnEntrys = (this->getViewStack()->getDisplay().getHeight()- 16) / 8 ;
```
Is `maximalDrawnEntrys` the max number of entries that fit on the screen? Might be good to clarify that

annawng commented on src/modm/ui/menu/choice_menu_impl.hpp:
```diff
@@ -11,43 +12,48 @@
+modm::ChoiceMenu<Allocator>::addEntry(const char* text,  bool *valuePtr, bool defaultValue)
 {
-	static uint16_t availableSpace = (getViewStack()->getDisplay().getWidth()-16)/6-6;
+	static uint16_t availableSpace = (this->getViewStack()->getDisplay().getWidth()-16)/6-6;
```
Not sure what the reasoning behind this math is or what `availableSpace` represents

MatthewMArnold replied:
me neither lol...this is stuff that we didn't write

annawng commented on src/modm/ui/menu/iabstract_view.hpp:
```diff
@@ -0,0 +1,122 @@
+		friend class ViewStack;
+
+		/**
+		 * @param stack pointer to the stack, the screen should be displayed on.
```
The comment says the stack pointer is a parameter but the only parameter of the constructor is the identifier.

MatthewMArnold replied:
Good catch

annawng commented on src/modm/ui/menu/iabstract_view.hpp:
```diff
@@ -0,0 +1,122 @@
+			// nothing to be done here
+		}
+
+	public:
+		const uint8_t identifier;
+		bool alive;
```
Why are there three different `public` sections in this class and do these values need to be public?

MatthewMArnold replied:
No idea...those values don't need to be public

annawng commented on src/modm/ui/menu/standard_menu_impl.hpp:
```diff
@@ -11,67 +12,68 @@
+modm::StandardMenu<Allocator>::draw()
 {
-	modm::GraphicDisplay* display = &getViewStack()->getDisplay();
+	typename modm::GraphicDisplay* display = &this->getViewStack()->getDisplay();
```
I'm confused what `typename` is used here for.

MatthewMArnold replied:
The gist is that since `getViewStack` returns a template object, we need to add typename before the type so the compiler knows this is actually a `modm::GraphicDisplay`.

Read a complete answer here: https://stackoverflow.com/questions/610245/where-and-why-do-i-have-to-put-the-template-and-typename-keywords

MatthewMArnold committed:
[menu] remove coupling between modm gui and modm menu view and menu

This is because the gui stuff requires colored menus
but the menu code should not have to have colored
menus. This wouldn't be an issue if we could dynamic
cast but on certain devices you can't do this.
